internal/commands/build.go

```diff
@@ -0,0 +1,259 @@
+package commands
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/docker/app/internal/packager"
+	"github.com/docker/distribution/reference"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/session/auth/authprovider"
+	"github.com/opencontainers/go-digest"
+	"github.com/sirupsen/logrus"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/docker/buildx/driver"
+
+	"github.com/docker/buildx/util/progress"
+
+	cnab "github.com/deislabs/cnab-go/driver"
+	"github.com/docker/buildx/bake"
+	"github.com/docker/buildx/build"
+	_ "github.com/docker/buildx/driver/docker" // required to get default driver registered, see driver/docker/factory.go:14
+	"github.com/docker/cli/cli"
+	"github.com/docker/cli/cli/command"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/spf13/cobra"
+)
+
+type buildOptions struct {
+	noCache  bool
+	progress string
+	pull     bool
+	tag      string
+	out		 string
+}
+
+func buildCmd(dockerCli command.Cli) *cobra.Command {
+	var opts buildOptions
+	cmd := &cobra.Command{
+		Use:     "build [APPLICATION]",
+		Short:   "Build service images for the application",
+		Example: `$ docker app build myapp.dockerapp`,
+		Args:    cli.RequiresRangeArgs(1, 1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			tag, err := runBuild(dockerCli, args[0], opts)
+			if err == nil {
+				fmt.Printf("Successfully build %s\n", tag.String())
+			}
+			return err
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.BoolVar(&opts.noCache, "no-cache", false, "Do not use cache when building the image")
+	flags.StringVar(&opts.progress, "progress", "auto", "Set type of progress output (auto, plain, tty). Use plain to show container output")
+	flags.BoolVar(&opts.pull, "pull", false, "Always attempt to pull a newer version of the image")
+	flags.StringVarP(&opts.out, "output", "o", "", "Dump generated bundle into a file")
+	flags.StringVarP(&opts.tag, "tag", "t", "", "Name and optionally a tag in the 'name:tag' format")
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
```

Comment:
Maybe should we find a way to split this method? To have some more identifiable chunks of work we can also unit test?

Reply:
Most of this method is about setting up parameters for buildx `Build`. I don't think there's any useful way to unit-test this. The last part about collecting digests and store bundle could be extracted into a simpler function, good point